lib/drivers/abbott/libreViewDriver.js

```diff
@@ -0,0 +1,356 @@
+      const mostRecent = sundial.applyTimezone(ts, config.timezone).toISOString();
+
+      Object.keys(payload.devices).forEach((key) => {
+        const device = payload.devices[key];
+        const events = device.data;
+
+        const postRecords = [];
+        payload.postRecords = postRecords;
+
+        config.tzoUtil = new TZOUtil(config.timezone, mostRecent, postRecords);
+        device.postRecords = postRecords;
+        config.builder.setDefaults({ deviceId: device.info.deviceId });
+
+        events.filter(elem => elem['Record Type'] === RECORD_TYPE_HISTORIC)
+          .forEach((record) => {
+            const cbg = config.builder.makeCBG()
+              .with_value(record.value)
+              .with_units(record.units)
+              .with_deviceTime(sundial.formatDeviceTime(record.jsDate))
+              .set('index', record.csvIndex);
+
+            config.tzoUtil.fillInUTCInfo(cbg, record.jsDate);
+
+            addOutOfRangeAnnotation(cbg, GLUCOSE_LO, GLUCOSE_HI, 1, 'bg');
+            postRecords.push(cbg.done());
+          });
+        events.filter(elem => (elem['Record Type'] === RECORD_TYPE_STRIP || elem['Record Type'] === RECORD_TYPE_SCAN))
+          .forEach((record) => {
+            const smbg = config.builder.makeSMBG()
+              .with_value(record.value)
+              .with_units(record.units)
+              .with_deviceTime(sundial.formatDeviceTime(record.jsDate))
+              .set('index', record.csvIndex);
+
+            if (record['Record Type'] === RECORD_TYPE_SCAN) {
+              smbg.with_subType('scanned');
+            }
+
+            addOutOfRangeAnnotation(smbg, GLUCOSE_LO[record.units], GLUCOSE_HI[record.units], 1, 'bg');
+            config.tzoUtil.fillInUTCInfo(smbg, record.jsDate);
+
+            postRecords.push(smbg.done());
+          });
+        events.filter(elem => elem['Record Type'] === RECORD_TYPE_KETONE)
+          .forEach((record) => {
+            const ketone = config.builder.makeBloodKetone()
+              .with_value(record.value)
+              .with_units(record.units)
+              .with_deviceTime(sundial.formatDeviceTime(record.jsDate))
+              .set('index', record.csvIndex);
+
+            addOutOfRangeAnnotation(ketone, KETONE_LO, KETONE_HI, 1, 'ketone');
+            config.tzoUtil.fillInUTCInfo(ketone, record.jsDate);
+            postRecords.push(ketone.done());
+          });
+      });
+      progress(100);
+      cb(null, payload);
+    },
+
+    uploadData: (progress, payload, cb) => {
+      progress(0);
+      payload.post_records = [];
+
+      async.eachSeries(Object.keys(payload.devices), (key, done) => {
+        const device = payload.devices[key];
+        const deviceRecords = device.postRecords;
+
+        const sessionInfo = {
+          deviceTags: ['bgm', 'cgm'],
+          deviceManufacturers: ['Abbott'],
+
+          deviceModel: key,
+          deviceSerialNumber: device.info.serialNumber,
+          deviceId: device.info.deviceId,
```

Comment:
Since we use the device ID for deduplication, it's probably not a good idea to have spaces in there (see https://developer.tidepool.org/data-model/device-data/common.html#deviceid for details).

Maybe just strip out the spaces, add the manufacturer and add a `-` between the name and serial number? Maybe a good idea to match the Libre driver, for example:

```js
deviceModel: "FreeStyle Libre"
deviceSerialNumber: "JCMW071-T1802"
deviceId: "AbbottFreeStyleLibre-JCMW071-T1802"
```

Reply:
from the CSV I have:
```
FreeStyle Libre 2,MBGAxxx-Gxxx,18-09-2021 16:33,6,,,,,,,,,,,,,,,
FreeStyle LibreLink,xxx-xxx-xxx-xxx-xxx,15-07-2021 02:14,0,139,,,,,,,,,,,,,,
FreeStyle Libre,JCMW071-T1802,23-07-2019 11:05,6,,,,,,,,,,,,,,,
```
so something like that:
```
deviceModel = FIELD_1
deviceSerialNumber = FIELD_2
deviceId = ("abott" + deviceModel + deviceSerialNumber).replaceAll(" ", "")
```

Reply:
Yeah, that looks good to me. I'd just add a `-` between device model and serial number so that it matches the existing Libre driver. Also, use `'` instead of `"` to keep the linter happy.

Reply:
updated following your suggestion

Reply:
> I'd just add a - between device model and serial number

> AbbottFreeStyleLibre-JCMW071-T1802

I took too long to notice that the first `-` was to be 'hardcoded' while the second one was part of the S/N 🙃

Reply:
No worries, thank you! I've fixed the last couple of linting issues. I'm going to send this off to our QA team!